File: ws01_plumbing/src/cpp01_topic/src/demo02_listener_str.cpp
/*
需求：订阅发布方发布的消息，并在终端输出
流程：
    1.包含头文件
    2.初始化 ROS2 客户端
    3.自定义节点类
        3-1.创建订阅方
        3-2.解析并输出数据
    4.调用spin函数，并传入节点对象指针
    5.资源释放
*/

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

// 3.自定义节点类;
class Listener : public rclcpp::Node
{
public:
    Listener() : Node("listener_node_cpp")
    {
        RCLCPP_INFO(this->get_logger(), "订阅方创建!");

        // 3-1.创建订阅方
        subcription_ = this->create_subscription<std_msgs::msg::String>("chartter", 10, std::bind(&Listener::do_cb, this, std::placeholders::_1));
        /*
        create_subscription
        模板:消息类型
        参数: 1.话题名称(和发布方保持一直) 2.QOS(队列长度) 3.回调函数
        返回值:订阅对象指针
        */
    }

private:
    void do_cb(const std_msgs::msg::String &msg)
    {
        // 3-2.解析并输出
        RCLCPP_INFO(this->get_logger(), "订阅到的消息为:%s", msg.data.c_str());
    }
    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subcription_;
};

int main(int argc, char const *argv[])
{
    // 2.初始化 ROS2 客户端
    rclcpp::init(argc, argv);

    // 4.调用spin函数,并传入节点对象指针
    rclcpp::spin(std::make_shared<Listener>());

    // 5.资源释放
    rclcpp::shutdown();
    return 0;
}

File: ws01_plumbing/src/py01_topic/py01_topic/demo02_listener_str.py
"""
需求：订阅发布方发布的消息，并在终端输出
流程：
    1.导包
    2.初始化 ROS2 客户端
    3.自定义节点类
        3-1.创建订阅方
        3-2.解析并输出数据
    4.调用spin函数,并传入节点对象
    5.资源释放
"""

# 1.导包
import rclpy
from rclpy.node import Node
from std_msgs.msg import String

# 3.自定义节点类
class Listener(Node):
    def __init__(self):
        super().__init__("listener_node_py")
        self.get_logger().info("订阅方创建了...")

        # 3-1.创建订阅方
        self.subscription = self.create_subscription(String, "chatter", self.do_cb, 10)

    def do_cb(self, msg):
        """
        回调函数 \n
        :param msg: 接收到的消息
        """

        # 3-2.解析并输出数据
        self.get_logger().info(f"订阅的数据{msg.data}")


def main():

    # 2.初始化
    rclpy.init()

    # 4.调用spin函数
    rclpy.spin(Listener())

    # 5.资源释放
    rclpy.shutdown()


if __name__ == "__main__":
    main()
